Adding Composer autoloader support to Magento

nevvermind opened the pull request.

adragus-inviqa committed:
Add Magento autoloader

adragus-inviqa committed:
Merge remote-tracking branch 'origin/develop' into feature/composer-autoloading

# Conflicts:
#	src/MagentoHackathon/Composer/Magento/ProjectConfig.php

adragus-inviqa committed:
Be explicit when failing + refactoring

svalaskevicius commented:
would it be better to change `index.php` instead of `Mage.php`?

its the file we change /usually/ when similar things are needed

nevvermind commented:
@svalaskevicius - yes, we thought about that, but what about api, shells or crons contexts?

jamescowie commented:
If we patch index.php then we also need to patch `cron.php` `shell.php` `api.php` and `get.php` as they are all the entry points into Magento. Where `Mage.php` invokes the autoloader for all and results in a single file being modified.

svalaskevicius commented:
true :)

jon-acker commented on src/MagentoHackathon/Composer/Magento/Patcher/Bootstrap.php:
```diff
@@ -24,86 +55,158 @@ private function getConfig()
+\$autoloaderPath = '{$this->getVendorAutoloaderPath()}';
+if (file_exists(\$autoloaderPath)) {
+    require_once \$autoloaderPath;
+}
```
if the file doesn't exist - shouldn't we throw an exception?, since if the patch was applied, the user will expect composers autoloader to have been loaded.

nevvermind replied:
The autoloader path if pretty much guaranteed (as much as you can with composer), because that path comes from here https://github.com/inviqa/magento-composer-installer/pull/3/files#diff-696ef84c63e9db98ae0c3ee424cfceb5R210 - aka directly from Composer's innards about the vendor-dir path.

jon-acker replied:
but there are conditions in which it won't work, e.g. running composer install on the host machine. 
To be clear, there's a conditional "file_exists", but you're not acting on its false path, so might as well not check file_exists  and just have "require_once \$autoloaderPath;"

nevvermind replied:
The `file_exists` is there as a runtime check, so that the **app** won't fail. But when the patched is applied, that's another thing - then we know for sure the file exists (otherwise Composer can't function).

Still, I agree. Let's throw an exception in there.

svalaskevicius replied:
just to check if its writing relative path to `vendors` or is it env specific?

adragus-inviqa committed:
Merge remote-tracking branch 'origin/develop' into feature/composer-autoloading

adragus-inviqa committed:
Merge branch 'develop' into feature/composer-autoloading

adragus-inviqa committed:
replace deprecated class

adragus-inviqa committed:
Change the path to the autoloader to be relative

For now, two paths are assumed: either two folders (mage root sibling with vendor) up or one (mage root as parent of mage root)
no need to check for the inclusion of Mage

adragus-inviqa committed:
add docs

adragus-inviqa committed:
fix json

adragus-inviqa committed:
new docs

adragus-inviqa committed:
add doc for vendor name changes

nevvermind commented:
This branch got merged into Cotya instead. Closing.
